### ISSUE TITLE
SuspendingConnection.inTransaction is running on multiple connections

### ISSUE THREAD

**Issue by oshai** [jasync-sql/jasync-sql#177]

**Comment by andy-yx-chen** [comment]
how can it be?

**Comment by oshai** [comment]
See the issue on gitter, I will give more details later: https://gitter.im/jasync-sql/support
I have an idea how fix it.

**Comment by leemorrisdev** [comment]
I raised this issue on gitter - adding some of the detail from there.

I'm creating a connection as follows:

`    
val connection = MySQLConnectionBuilder.createConnectionPool(databaseUrl) {
        maxActiveConnections = 8
        maxIdleTime = 60000
        applicationName = "ktor-test"
    }
`

I have a repository that is based on coroutines, and uses the following to create a transaction and do a few inserts:

`
suspend fun create(parcel: Parcel): Long = withContext(Dispatchers.IO) {

        connection.asSuspending.inTransaction { txn ->

            val pickupAddressId =  createAddress(txn, parcel.pickupAddress)
            val deliveryAddressId = createAddress(txn, parcel.deliveryAddress)
            val recipientId =  createContact(txn, parcel.recipient)

            createParcel(txn, parcel, pickupAddressId, deliveryAddressId, recipientId)
        }
    }
`

Each call just uses the txn object to do an insert, each function uses suspend.

The logs show the following for a run of this which produced a broken set of records in the DB:

`
2020-01-03 09:29:48.064 [DefaultDispatcher-worker-2] TRACE c.g.j.sql.db.mysql.MySQLConnection - <mysql-connection-4> sendPreparedStatement() - PreparedStatementParams(query=insert into address (address_1, address_2, address_3, city, postcode)
values (?, ?, ?, ?, ?), values=[3 Northridge Terrace, null, null, Sutysky, AA], release=false)

2020-01-03 09:29:48.066 [DefaultDispatcher-worker-4] TRACE c.g.j.sql.db.mysql.MySQLConnection - <mysql-connection-5> sendPreparedStatement() - PreparedStatementParams(query=insert into address (address_1, address_2, address_3, city, postcode)
values (?, ?, ?, ?, ?), values=[5 Northridge Terrace, null, null, Sutysky, BB], release=false)

2020-01-03 09:29:48.067 [DefaultDispatcher-worker-9] TRACE c.g.j.sql.db.mysql.MySQLConnection - <mysql-connection-2> sendPreparedStatement() - PreparedStatementParams(query=insert into contact(first_name, last_name, phone_number)
values (?, ?, ?), values=[Sax, Delepine, 142-396-0420], release=false)

2020-01-03 09:29:48.069 [DefaultDispatcher-worker-11] TRACE c.g.j.sql.db.mysql.MySQLConnection - <mysql-connection-1> sendPreparedStatement() - PreparedStatementParams(query=insert into parcel(carrier_ref, carrier_code, pickup_at, pickup_address_id, delivery_address_id, recipient_id, pieces)
values (?, ?, ?, ?, ?, ?, ?), values=[00000019, XXX, Wed Oct 03 01:00:00 BST 2018, 39, 40, 20, 1], release=false)
`

As you can see it appears my transaction is running across multiple connections so records are being inserted at (effectively) random when that connection eventually gets a COMMIT executed.

If required I can extract a sample project that reproduces this, I used gatling to generate enough load to cause the problem.

**Comment by oshai** [comment]
The issue is that inTransaction is not implemented correctly for suspending connection:
```
override suspend fun <A> inTransaction(f: suspend (SuspendingConnection) -> A): A {
        this.sendQuery("BEGIN")
        try {
            val result = f(this)
            this.sendQuery("COMMIT")
            return result
        } catch (e: Throwable) {
            this.sendQuery("ROLLBACK")
            throw e
        }
    }
```
calling `this.sendQuery()` on a pool will bring different connection each time.

**Comment by oshai** [comment]
Fixed in 1.0.13

**Comment by oshai** [comment]
@leemorrisdev Thanks for reporting and investigating the issue!
